[selectors] Add tests for :focus-visible in the default UA style sheet

mrego opened the pull request:
This patch adds 2 new tests to verify that the default UA style sheet
uses `:focus-visible { outline: auto; }`.
See: whatwg/html#6256 & w3c/csswg-drafts#4278

* focus-visible-017.html:
  This test checks that when you focus an element via script,
  it show a focus ring with `outline-style: auto`.
  Currently Chromium passes this test,
  because despite they don't use `:focus-visible` in the UA stylesheet,
  it's painting an auto style outline when an element is focused.
  However Firefox fails it, because even when it uses `:-moz-focusring`
  (the equivalent to `:focus-visible`) in the UA stylesheet,
  it uses dotted style for the outline.
  WebKit doesn't support `:focus-visible` yet an it fails,
  thought it's painting an auto style outline
  (the test is specifically checking for `:focus-visible` support).

* focus-visible-018.html:
  This test checks that when you click an element to focus it,
  it doesn't show any kind of focus ring.
  Currently Firefox passes this test, by Chromium fails it
  because Chromium is using `:focus` on the default UA stylesheet
  and is adding an outline on the element, despite it doesn't match
  `:focus-visible` (see https://crbug.com/1162070).

mrego committed:
[selectors] Add tests for :focus-visible in the default UA style sheet

This patch adds 2 new tests to verify that the default UA style sheet
uses `:focus-visible { outline: auto; }`.
See: whatwg/html#6256 & w3c/csswg-drafts#4278

* focus-visible-017.html:
  This test checks that when you focus an element via script,
  it show a focus ring with `outline-style: auto`.
  Currently Chromium passes this test,
  because despite they don't use `:focus-visible` in the UA stylesheet,
  it's painting an auto style outline when an element is focused.
  However Firefox fails it, because even when it uses `:-moz-focusring`
  (the equivalent to `:focus-visible`) in the UA stylesheet,
  it uses dotted style for the outline.
  WebKit doesn't support `:focus-visible` yet an it fails,
  thought it's painting an auto style outline
  (the test is specifically checking for `:focus-visible` support).

* focus-visible-018.html:
  This test checks that when you click an element to focus it,
  it doesn't show any kind of focus ring.
  Currently Firefox passes this test, by Chromium fails it
  because Chromium is using `:focus` on the default UA stylesheet
  and is adding an outline on the element, despite it doesn't match
  `:focus-visible` (see https://crbug.com/1162070).

emilio commented:
Actually, hold on, I think this might need HTML spec changes.

alice commented on css/selectors/focus-visible-017.html:
```diff
@@ -0,0 +1,36 @@
+<script>
+  async_test(function(t) {
+    target.addEventListener("focus", t.step_func(function() {
+      assert_equals(getComputedStyle(target).outlineStyle, "auto", `outline-style for ${target.tagName}#${target.id} should be auto`);
```
I think this test is a little misleading. The spec states:

> If the active element matches `:focus-visible`, and a script causes focus to move elsewhere, the newly focused element should match `:focus-visible`.
> 
> Conversely, if the active element does not match `:focus-visible`, and a script causes focus to move elsewhere, the newly focused element should **not** match `:focus-visible`.

This is tested by https://github.com/web-platform-tests/wpt/blob/master/css/selectors/focus-visible-005.html and https://github.com/web-platform-tests/wpt/blob/master/css/selectors/focus-visible-008.html respectively.

What is this test actually testing?

mrego replied:
This 2 tests are to cover this part of the spec:
> User agents should also use :focus-visible to specify the default focus style, so that authors using :focus-visible will not also need to disable the default :focus style.

Right now Chromium is not using `:focus-visible` yet for that, so when you use the mouse click to focus an element, it always shows a focus ring (see https://crbug.com/1162070).

About this particular test, this is testing that the `outline-style` is exactly `auto` in this case. @domenic asked for this test at whatwg/html#6256. That PR to HTML spec is adding: `:focus-visible { outline: auto; }` to the default UA style sheet.
Chromium passes this test right now, even when it currently uses `:focus` for that. And Firefox fails it because even when it uses `:-moz-focusring` (which is equivalent to `:focus-visible`), it sets a `dotted` which won't be following the HTML spec.

alice replied:
Programmatic focus doesn't always cause `:focus-visible` to match, so maybe we should test both with and without `:focus-visible` matching being triggered?

For example, make a separate element (say, a `<button>`, called something like `initial`), which has a click listener which moves focus to `target`. 

Then, first send `initial` a mouse click, so that `target` *doesn't* match `:focus-visible`, and check that it has no outline.

Next, move focus back to `initial`, and send it an `enter` keypress to trigger the click handler (via a synthetic click). Then, `target` *should* match `:focus-visible`, and thus *should* have an outline.

mrego replied:
That kind of interactions are already tested in `focus-visible-{014,015,016}.html` tests. Here I was just looking for a case when `:focus-visible` matches to check the default UA style.

I can change this test to send a TAB key to do the focus, or keep it as is with a click, but reword it so it's way clearer what we're testing here. I have followed the latter approach in the new patch.

If you want to do what you suggest here, probably we can just have a single test ( `013`) that checks both things, the default UA stylesheet when `:focus-visible` doesn't match and then when `:focus-visible` matches. Let me know if you prefer that option.

alice commented on css/selectors/focus-visible-013.html:
```diff
@@ -0,0 +1,46 @@
+  }
+</style>
+
+<p>This test checks that using the mouse to focus an element does not trigger <code>:focus-visible</code> matching, and the element doesn't show any focus ring.</p>
```
The first part of this is already well covered by existing tests; what exactly is this test checking?

mrego replied:
This test current fails in Chromium and passes in Firefox. So it's not totally covered by other tests.

This checks that when you focus via mouse click, the element doesn't get a focus ring. As it doesn't match `:focus-visble` it shouldn't get any outline by default. The default UA style sheet rule is going to be changed from `:focus { outline: auto; }` to `:focus-visible { outline: auto; }`, so Chromium shouldn't show any outline by default when `:focus-visible` doesn't match, like in this test.

alice replied:
Got it, thanks for the explanation.

That being the case, I think we don't need the custom styles (red/green background), since the selector matching part is already covered by other tests.

Instead, we could reword the test as something like "... by default, using the mouse to focus a generic element does not show a focus ring". Then, we could just check the outline style on focus. If the selector isn't supported, or isn't the UA default, this will fail.

mrego replied:
I agree that the customs styles are not needed for the test, but I think they're useful as it's covered in other tests. But I don't think it's bad to have them, as that makes easier to identify a failure if `:focus-visible` is not supported at all.

I've reworded things as suggested but I'll keep the background colors unless you see a big problem with them.

alice replied:
No problem if you strongly prefer to keep the background colours, but why not just have something like the following?

```css
.warning {
  display: none;
}

:supports not (:focus-visible) {
  .warning {
    display: initial;
  }
}
```

```html
<div class="warning">
<code>:focus-visible</code> is not supported. This test will not pass.
</div>
```

mrego replied:
The problem with that warning is that it won't change anything on the test, and maybe the script tests could be true without any `:focus-visible` support at all. Probably not this particular test but it might happen in other tests.

I want that a browser that doesn't support `:focus-visible` at all, fail on the automatic tests. So using this kind of red outlines/backgrounds is usually useful for both the automatic testing and manual testing.

Also I've added them in most of the test on the suite in previous patches (before that some tests were passing in WebKit without any kind of implementation), so maybe if we want to change that we could do for all of them together in follow-up patches.

alice replied:
Ok, maybe just add a comment explaining what the background colours are testing then, so we don't accidentally remove them later.

alice commented on css/selectors/focus-visible-013.html:
```diff
@@ -0,0 +1,46 @@
+<!DOCTYPE html>
+<meta charset="utf-8" />
+<title>CSS Test (Selectors): Mouse focus does not match :focus-visible</title>
```
Would also need to update this. Something like "mouse focus does not cause an outline to be shown by default".

alice reviewed (commented):
Thanks for the response, and for taking this work on!

mrego commented:
Thanks for the review!

I've uploaded a new version rewording the test title, expectations and so on. Let me know what do you think about the new version.

alice commented on css/selectors/focus-visible-017.html:
```diff
@@ -0,0 +1,36 @@
+      assert_equals(getComputedStyle(target).backgroundColor, "rgb(0, 255, 0)", `backgroundColor for ${target.tagName}#${target.id} should be lime`);
+      t.done();
+    }));
+    target.focus();
```
To clarify my earlier comment: this line only triggers `:focus-visible` matching coincidentally, because we match `:focus-visible` by default when the page loads. Depending on that kind of unspecified behaviour makes me anxious. 

I explained in my earlier comment how to ensure that `.focus()` causes the newly focused element to match `:focus-visible`.

mrego replied:
I didn't realize this was not specified behavior... isn't that what https://wpt.fyi/results/css/selectors/focus-visible-010.html is testing? Or am I missing something else?

I understand now why you asked for adding more elements.

alice replied:
Oh haha, I guess I forgot about that! Thanks for reminding me; in that case this is fine.

alice replied:
(Might be worth adding a comment explaining that, in case I forget again.)

mrego replied:
TBH I cannot find this behavior defined on [the spec](https://drafts.csswg.org/selectors-4/#the-focus-visible-pseudo), so it's `focus-visible-010.html` that is wrong, or is the spec not updated?

The spec says:
* If the active element matches :focus-visible, and a script causes focus to move elsewhere, the newly focused element should match :focus-visible.
* Conversely, if the active element does not match :focus-visible, and a script causes focus to move elsewhere, the newly focused element should not match :focus-visible.

But doesn't mention the initial focus, when there's no active element before the focus. WDYT?

alice replied:
I suspect we snuck it into tests without adding it to the spec. Do you think we should add it to the list of heuristics as well?

mrego replied:
I think it'd be nice adding it yeah, I can fill a CSSWG issue to do that. As Chromium and Firefox are already passing `focus-visible-010.html` so it doesn't need implementation changes.

alice replied:
Thank you!

mrego replied:
Done: https://github.com/w3c/csswg-drafts/issues/5885

Thank you for the thorough reviews and all the feedback on these tests!

mrego commented:
I've uploaded a new version of the 2 tests, and also updated the initial comment and commit message to explain them.

I've removed the background colors and instead added a `warning` message when user is checking manually this test, and used `test_valid_selector(':focus-visible');` to verify `:focus-visible` support when this is run automatically.

I've added manual instructions to both tests too.

I hope the new version of the tests and description in the commit message are good enough.

Thanks for the reviews.

mrego commented:
BTW, the [HTML spec change](https://github.com/whatwg/html/pull/6256) has been merged, so this should be ready to be merged once @alice does the last review pass. Thanks!

alice reviewed (approved):
Thanks again for your patience and work on this!

mrego commented:
Thanks for the reviews.
